src/Simplex/Chat.hs

```diff
@@ -7460,6 +7487,35 @@ closeFileHandle fileId files = do
   h_ <- atomically . stateTVar fs $ \m -> (M.lookup fileId m, M.delete fileId m)
   liftIO $ mapM_ hClose h_ `catchAll_` pure ()
 
+tagGroupChat' :: DB.Connection -> User -> GroupId -> ChatTagId -> ExceptT StoreError IO ChatResponse
+tagGroupChat' db user gId ctId = do
+  liftIO $ tagGroupChat db gId ctId
+  (allTags, chatTags) <- updatedGroupChatTags db user gId
+  pure $ CRChatTagged user allTags chatTags
+
+updatedGroupChatTags :: DB.Connection -> User -> GroupId -> ExceptT StoreError IO ([ChatTag], [ChatTag])
+updatedGroupChatTags db user gId = do
+  allTags <- liftIO $ getUserChatTags db user
+  let (groupTags, otherTags) = partition (\ChatTag {groupId} -> groupId == Just gId) allTags
+  pure (groupTags, otherTags)
+
+deleteChatTagIfEmpty :: DB.Connection -> User -> ChatTagId -> ExceptT StoreError IO ()
+deleteChatTagIfEmpty db user ctId = do
+  tagChatsCount <- liftIO $ getTagChatsCount db ctId
+  when (tagChatsCount == 0) $ liftIO $ deleteChatTag db user ctId
+
+tagDirectChat' :: DB.Connection -> User -> ContactId -> ChatTagId -> ExceptT StoreError IO ChatResponse
+tagDirectChat' db user cId ctId = do
+  liftIO $ tagDirectChat db cId ctId
+  (allTags, chatTags) <- updatedDirectChatTags db user cId
+  pure $ CRChatTagged user allTags chatTags
+
+updatedDirectChatTags :: DB.Connection -> User -> ContactId -> ExceptT StoreError IO ([ChatTag], [ChatTag])
+updatedDirectChatTags db user cId = do
+  allTags <- liftIO $ getUserChatTags db user
+  let (directTags, otherTags) = partition (\ChatTag {contactId} -> contactId == Just cId) allTags
+  pure (directTags, otherTags)
```

Comment:
Put them in Store folder together with queries and return tuples from them. The current boundary (and the other functions) might be unnecessary or should be internal to store.